packagekit: Automatic apt updates via `unattended-upgrades`

mchlwbr opened the pull request:
This is currently based on cockpit version 337 as we ([Team LinOTP](https://github.com/linotp)) develop/run on [bookworm-backports](https://packages.debian.org/de/bookworm-backports/cockpit) and resolves #15269.

I will happily rebase on a branch of your choice to get the feature to `bookworm-backports`.

#18595 seems to be stale and instead of parsing the config files on our own, we rely on `apt-config dump` doing the heavy lifting.

mchlwbr committed:
packagekit: Automatic apt updates via `unattended-upgrades`

jelly commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -337,6 +337,144 @@ class Dnf5Impl extends ImplBase {
+            } else {
+                if (timerOutput.indexOf("InactiveSec=1d\n") >= 0)
+                    this.day = this.time = "";
+                else if (this.installed)
```
This should be unreachable, we return directly if it is not installed. (same issue is in the dnf5 implementation)

jelly commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -337,6 +337,144 @@ class Dnf5Impl extends ImplBase {
+                script += '        "origin=Debian,codename=${distro_codename}-security,label=Debian-Security";\n';
+            } else {
+                // All updates (security + regular)
+                script += '        "origin=Debian,codename=${distro_codename},label=Debian";\n';
```
Is this correct? Reading the periodic updates [wiki](https://wiki.debian.org/PeriodicUpdates), it says the default is only security updates and that is:

```
Unattended-Upgrade::Origins-Pattern {
        // Codename based matching:
        // This will follow the migration of a release through different
        // archives (e.g. from testing to stable and later oldstable).
        // Software will be the latest available for the named release,
        // but the Debian release itself will not be automatically upgraded.
//      "origin=Debian,codename=${distro_codename}-updates";
//      "origin=Debian,codename=${distro_codename}-proposed-updates";
        "origin=Debian,codename=${distro_codename},label=Debian";
        "origin=Debian,codename=${distro_codename},label=Debian-Security";
        "origin=Debian,codename=${distro_codename}-security,label=Debian-Security";
//      "o=Debian Backports,n=${distro_codename}-backports,l=Debian Backports";

        // Archive or Suite based matching:
        // Note that this will silently match a different release after
        // migration to the specified archive (e.g. testing becomes the
        // new stable).
//      "o=Debian,a=stable";
//      "o=Debian,a=stable-updates";
//      "o=Debian,a=proposed-updates";
//      "o=Debian Backports,a=stable-backports,l=Debian Backports";
};
```

So I think if you want "all updates" you also have to enable:
```
"origin=Debian,codename=${distro_codename}-updates";
```

And we also have to handle if a user has enabled backports I suppose and enable those updates.

jelly commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -337,6 +337,144 @@ class Dnf5Impl extends ImplBase {
+            // Determine upgrades type based on Origins-Pattern
+            const regularUpgrades = aptConfigOutput.includes('Unattended-Upgrade::Origins-Pattern:: "origin=Debian,codename=${distro_codename},label=Debian"');
+            const securityUpgrades = aptConfigOutput.includes('Unattended-Upgrade::Origins-Pattern:: "origin=Debian,codename=${distro_codename},label=Debian-Security"') ||
+                                   aptConfigOutput.includes('Unattended-Upgrade::Origins-Pattern:: "origin=Debian,codename=${distro_codename}-security,label=Debian-Security"');
```
These are very long lines and not to readable, I have some questions about the difference between all / security updates below. Hopefully we can simplify things after that is cleared up.

mchlwbr replied:
I've suggested to set `origins=*` for `all` updates below.

I will think about how the parsing of a given config can be refactored.

jelly commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -337,6 +337,144 @@ class Dnf5Impl extends ImplBase {
+                script += "printf '[Timer]\\nOnCalendar=\\nOnCalendar=" + day + " " + time + "\\n' > " + timerConf + "; ";
+                script += "systemctl daemon-reload; ";
+            }
+        }
```
This could become more generic and shared with the Dnf5Impl, but can be left as a follow up.

jelly reviewed (changes requested):
Thank you very much for picking this up, this has been a long awaited feature from the community. Sorry for the long delay for an initial review.

mchlwbr commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -337,6 +337,144 @@ class Dnf5Impl extends ImplBase {
+            const timerOutput = await cockpit.script(
+                "set -eu; " +
+                "if systemctl --quiet is-enabled apt-daily-upgrade.timer 2>/dev/null; then echo enabled; fi; " +
+                "systemctl cat apt-daily-upgrade.timer 2>/dev/null | grep '^OnCalendar=' | tail -n1 || true; ",
```
Correct. I will validate it too.

mchlwbr reviewed (commented):
Thanks for your review and sorry about the delay - I was on vacation :)

I will implement your suggestions.

Can you give me the OK for setting `"origin=*";` when the user selects `all` updates?

jelly reviewed (commented):
Thanks for the fixes!

cockpituous commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -337,6 +336,142 @@ class Dnf5Impl extends ImplBase {
+            return;
+        }
+
+        try {
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22515-acb2e3fc-20251125-163641-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22515-acb2e3fc-20251125-163641-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

mchlwbr commented on test/verify/check-packagekit:
```diff
@@ -1620,7 +1667,7 @@ class TestAutoUpdates(NoSubManCase):
         assertTimer(None)
 
-        if "dnf" in self.backend:
+        if "dnf" in self.backend or self.backend == "apt":
```
Sadly this doesn't work, as we have to account for both backend being `dnf4` and `dnf5`: https://github.com/cockpit-project/cockpit/blob/acb2e3fcb424a7d6b5dfb93cc35a665102da6fc0/test/common/packagelib.py#L42

So I tried to conform with the syntax in the other  tests.

mchlwbr commented:
Hi @jelly 
Thanks to the tests, I just noticed that the current implementation does not support ubuntu.

I'll get onto that and also extend the tests.

jelly commented on test/verify/check-packagekit:
```diff
@@ -1460,14 +1466,16 @@ class TestAutoUpdates(NoSubManCase):
         super().setUp()
-        # not implemented for apt yet, only dnf
-        self.supported_backend = "dnf" in self.backend
+        self.supported_backend = "dnf" in self.backend or self.backend == "apt"
```
One missing converted test is `testWithAvailableUpdates` it now fails with `NotImplementedError`.

jelly commented:
> Hi @jelly Thanks to the tests, I just noticed that the current implementation does not support ubuntu.
> 
> I'll get onto that and also extend the tests.

Thanks! Feel free to ask for help

cockpituous commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -90,6 +90,29 @@ class ImplBase {
+        let script = "";
+
+        if (time !== null || day !== null) {
+            if (day === "" && time === "") {
```
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22515-acb2e3fc-20251126-085322-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22515-acb2e3fc-20251126-085322-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

mchlwbr committed:
packagekit: Factorize timer drop-in creation script

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22515-7fd3f4c9-20260204-075640-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

martinpitt reviewed (changes requested):
Thanks! I'm afraid we'll have to discuss the implementation a bit. I did not yet look at the tests, and they also fail everywhere: [debian-testing](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22515-7fd3f4c9-20260204-075621-debian-testing-expensive/log.html), [debian-trixie](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22515-7fd3f4c9-20260204-075751-debian-trixie-expensive/log.html), [ubuntu-2404](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22515-7fd3f4c9-20260204-074845-ubuntu-2404-expensive/log.html), [ubuntu-stable](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22515-7fd3f4c9-20260204-074849-ubuntu-stable-expensive/log.html).

I will however do another force-push for the split out "packagekit: Factorize timer drop-in creation script" just to avoid losing that.

Thanks!

mchlwbr committed:
WIP: refactor: apt auto-updates

- Add RandomizedDelaySec=0 to systemd timer drop-ins to ensure
  updates run at the configured time.

- Use default unattended-upgrades config for security updates instead of
  duplicating security origin patterns. The "security" mode now relies
  on the default 50-unattended-upgrades config, which enables security
  updates for both Debain and Ubuntu, by *not* overriding
  Unattended-Upgrade::Allowed-Origins (Ubuntu) or Origins-Pattern
  (Debian), rather than explicitly listing all security sources.

- Detect custom configurations by checking for #clear directives in
  apt.conf.d/. If origins are cleared without our wildcard pattern,
  treat as unsupported custom config.

mchlwbr commented:
Hi @martinpitt 

I pushed a work-in-progress refactoring, that tackles your raised issues.

Some remarks:

- For detecting wether security updates are enabled I opted to check if the origins where cleared. If they are not cleared, we can assume that all security origins are included in the config as they come per default on both debian and ubuntu.
- I kept the suffix `-cockpit` for the customConfig, because I dont want to (potentially) overwrite existing configs of users
- I kept the distro detection to be able to set the correct keys/patterns on debian and ubuntu. (Maybe I misunderstood, but in https://github.com/cockpit-project/cockpit/pull/22515#discussion_r2762774057 it seems like you wanted to drop that)
- I did not yet rework the tests

Would you be up for a remote meeting next week if theres the need? Is your mail still martin@piware.de?

Best Regards!
Michi
